bootstrap/lib/cache.js

```diff
@@ -37,31 +50,58 @@ const withCache = (execute, options) => {
   // Algorithm we use to derive entry key
   const hashOptions = {
     algorithm: 'sha1',
+
     encoding: 'hex',
     excludeKeys: (key) => options.ignoredKeys.includes(key),
   }
+  const _getKey = (data) => hash(data, hashOptions)
+  const _getCoalescingKey = (key) => hash(`${options.requestCoalescingGroup}--${key}`, hashOptions)
 
   const _getMaxAge = (data) => {
     if (!data || !data.data) return options.maxAge
     if (isNaN(data.data.maxAge)) return options.maxAge
     return data.data.maxAge || options.maxAge
   }
 
-  return (data, callback) => {
-    const key = hash(data, hashOptions)
+  const _getWithCoalescing = async (key, coalescingKey, retries = 5, ms = 100) => {
+    if (retries === 0) return null
+    const entry = cache.get(key)
+    if (entry) {
+      logger.debug(`Request coalescing: Success`)
+      return entry
+    }
+    const inFlight = cache.get(coalescingKey)
+    if (!inFlight) return null
+    // Capped linear back-off: 100, 200, 400, 800, 1000..
+    ms = Math.min(ms * 2, 1000)
+    logger.debug(`Request coalescing: Waiting ${ms}ms`)
+    await delay(ms)
+    return await _getWithCoalescing(key, coalescingKey, retries - 1, ms)
+  }
+
+  return async (data, callback) => {
+    const key = _getKey(data)
+    const coalescingKey = _getCoalescingKey(key)
     const maxAge = _getMaxAge(data)
     // Add successful result to cache on callback
     const _cacheAndCallback = (statusCode, data) => {
       if (statusCode === 200) {
         const entry = { statusCode, data, maxAge }
         cache.set(key, entry, maxAge)
         logger.debug(`Cache set: ${key} => `, entry)
+
+        // Trigger request coalescing by removing in-flight mark
+        cache.del(coalescingKey)
```

Comment:
I would make sure the coalescingKey has a TTL so resource starvation doesn't occur from the following scenario:
- Process A grabs lock and then crashes
- Processes B, C, D, .... etc all have to wait the timeout length of the linear retry, until it finally returns null anyway